cro-clib/src/lib.rs

```diff
@@ -0,0 +1,263 @@
+use chain_core::init::network::init_chain_id;
+use client_common::error::ErrorKind;
+use client_common::storage::MemoryStorage;
+use client_common::SecKey;
+use client_core::types::WalletKind;
+use client_core::wallet::{DefaultWalletClient, WalletClient};
+use client_core::HDSeed;
+use client_core::Mnemonic;
+use secstr::SecUtf8;
+use std::ffi::CStr;
+use std::os::raw::c_char;
+use std::str::FromStr;
+pub const MAX_LENGTH: u32 = 1024;
+pub const MAX_STRING_LENGTH: u32 = 512;
+#[repr(C)]
+#[derive(Copy, Clone)]
+pub struct Buf {
+    pub buf: [::std::os::raw::c_uchar; 512usize],
+    pub length: ::std::os::raw::c_int,
+}
+
+#[repr(C)]
+#[derive(Copy, Clone)]
+pub struct ApiResult {
+    pub error: ::std::os::raw::c_int,
+    pub value: Buf,
+}
+
+#[repr(C)]
+#[derive(Copy, Clone)]
+pub struct HDWallet {
+    pub error: ::std::os::raw::c_int,
+    pub name: Buf,
+    pub value: Buf,
+    pub mnemonics: Buf,
+    pub viewkey: Buf,
+    pub seed: Buf,
+    pub enckey: Buf,
+}
+
+#[repr(C)]
+#[derive(Copy, Clone)]
+pub struct ApiContext {
+    pub error: ::std::os::raw::c_int,
+    pub chain_id: Buf,
+    pub server_url: Buf,
+    pub storage_folder: Buf,
+}
+
+static mut G_DB: Option<MemoryStorage> = None;
+
+/// # Safety
+/// this function should not be called inside rust
+unsafe fn get_string(src: *const c_char) -> String {
+    CStr::from_ptr(src).to_string_lossy().into_owned()
+}
+
+fn copy_string(src: &str, dst: &mut [u8]) {
+    dst[..src.len()].clone_from_slice(&src.as_bytes()[..src.len()]);
+    dst[src.len()] = 0;
+}
+
+fn get_string_from_buf(src: &Buf) -> String {
+    std::str::from_utf8(&src.buf[0..src.length as usize])
+        .expect("convert to string")
+        .to_string()
+}
+
+fn get_buf(src: &str) -> Buf {
+    let mut ret = Buf {
+        buf: [0; MAX_STRING_LENGTH as usize],
+        length: 0,
+    };
+    copy_string(src, &mut ret.buf);
+    ret.length = src.len() as i32;
+    ret
+}
+
+fn get_buf_vec(src: &[u8]) -> Buf {
+    let mut ret = Buf {
+        buf: [0; MAX_STRING_LENGTH as usize],
+        length: 0,
+    };
+    ret.buf[..src.len()].clone_from_slice(&src[..]);
+    ret.length = src.len() as i32;
+    ret
+}
```

Comment:
these are very error-prone and dangerous.

Perhaps when different types are used for inputs-outputs, instead of one fixed array.

See the previous feedback:

https://github.com/crypto-com/chain/pull/891#discussion_r366741011
https://github.com/crypto-com/chain/pull/891#discussion_r366738727

Let the caller manage allocation in any way they like, so just accept `*mut ...` + length... if the length is not enough, return an error instead of silently copying only a part of data

Reply:
because main data will be rust object and handled in rust side.
that code will be removed.